src/inmanta/deploy/work.py

```diff
@@ -0,0 +1,424 @@
+    #########################
+    # Mapping implementation#
+    #########################
+
+    def __getitem__(self, key: Task) -> PrioritizedTask[Task]:
+        return self._tasks_by_resource.get(key.resource, {})[key].task
+
+    def __iter__(self) -> Iterator[Task]:
+        return itertools.chain.from_iterable(self._tasks_by_resource.values())
+
+    def __len__(self) -> int:
+        return sum(len(items) for items in self._tasks_by_resource.values())
+
+
+@dataclass
+class BlockedDeploy:
+    """
+    Deploy task that is blocked on one or more of its dependencies (subset of its requires relation).
+    """
+    task: PrioritizedTask[Deploy]
+    blocked_on: set[ResourceIdStr]
+
+
+class ScheduledWork:
+    """
+    Collection of deploy tasks that should be executed. Manages a priority queue of ready-to-execute tasks per agent
+    (AgentQueues), as well as an unordered collection of waiting deploy tasks (across agents). Maintains the following
+invariants in between method calls:
+    - Each task exists at most once in the scheduled work. It is however possible for an identical task to be executing.
+    - A resource's requires relation is enforced within the set of scheduled work, i.e. for any two deploy tasks in the
+        scheduled work, they will be processed in requires order. Only direct requires are considered. It is the responsibility
+        of the scheduler to include in-between resources as scheduled work if/when transitive requires ordering is desired.
+
+    Expects to be informed by the scheduler of deploy requests and/or state changes through update_state() and
+    delete_resource().
+
+    Expects to be informed by task runners of finished tasks through notify_provides().
+
+    :param requires: Live, read-only view on requires-provides mapping for the latest model state.
+    """
+
+    def __init__(self, requires: state.RequiresProvidesMapping) -> None:
+        self.requires: state.RequiresProvidesMapping = requires
+        self.agent_queues: AgentQueues = AgentQueues()
+        self.waiting: dict[ResourceIdStr, BlockedDeploy] = {}
+
+    # TODO: name
+    def update_state(
+        self,
+        *,
+        ensure_scheduled: Set[ResourceIdStr],
+        running_deploys: Set[ResourceIdStr],
+        added_requires: Optional[Mapping[ResourceIdStr, Set[ResourceIdStr]]] = None,
+        dropped_requires: Optional[Mapping[ResourceIdStr, Set[ResourceIdStr]]] = None,
+    ) -> None:
+        """
+        Update the scheduled work state to reflect the new model state and scheduler intent. May defer tasks that were
+previously considered ready to execute if any of their dependencies are added to the scheduled work.
+
+        :param ensure_scheduled: Set of resources that should be deployed. Adds a deploy task to the scheduled work for each
+            of these, unless it is already scheduled.
+        :param running_deploys: Set of resources for which a deploy is currently running, for the latest desired state, i.e.
```

Comment:
this I don't understand

Reply:
It was a bit of a last-minute addition, may need more refinement. I'll see what I can do in #8008 and if nothing else I'll clarify it there.